torch/csrc/api/include/torch/optim/serialize.h

```diff
@@ -214,13 +214,21 @@ void serialize(
   torch::optim::serialize<OptimizerName##ParamState, OptimizerName##Options>(archive, self)
 
 #define _TORCH_OPTIM_SERIALIZE_TORCH_ARG(name) \
-  archive.write(#name, IValue(name()))
+  archive.write(#name, IValue(name())); \
+
+// serialize only if the tensor is defined
+#define _TORCH_OPTIM_SERIALIZE_TORCH_ARG_IF_DEFINED(name) \
+{ \
+  if(name().defined()) \
+    _TORCH_OPTIM_SERIALIZE_TORCH_ARG(name); \
+} \
```

Comment:
It's a little sad that we need two macros `_TORCH_OPTIM_SERIALIZE_TORCH_ARG_IF_DEFINED` and `_TORCH_OPTIM_SERIALIZE_TORCH_ARG` to support serializing Tensor and non-Tensor objects. But actually we can merge them into one:
```cpp
#define _TORCH_OPTIM_SERIALIZE_TORCH_ARG(name) \
{ \
  auto ivalue = torch::IValue(name()); \
  if (!(ivalue.isTensor() && ivalue.unsafeToTensorImpl() == at::UndefinedTensorImpl::singleton())) { \
    archive.write(#name, ivalue); \
  } \
}
```